File: app/src/main/java/com/zaze/demo/component/wifi/ui/NetworkStatsActivity.java
package com.zaze.demo.component.wifi.ui;

import android.os.Bundle;

import com.zaze.common.base.mvp.BaseMvpActivity;
import com.zaze.demo.R;
import com.zaze.demo.component.wifi.adapter.NetworkStatsAdapter;
import com.zaze.demo.component.wifi.presenter.NetworkStatsPresenter;
import com.zaze.demo.component.wifi.presenter.impl.NetworkStatsPresenterImpl;
import com.zaze.demo.component.wifi.view.NetworkStatsView;
import com.zaze.demo.debug.NetTrafficStats;

import java.util.Collection;

import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout;


/**
 * Description :
 *
 * @author : zaze
 * @version : 2017-12-22 04:36 1.0
 */
public class NetworkStatsActivity extends BaseMvpActivity<NetworkStatsView, NetworkStatsPresenter> implements NetworkStatsView {

    private RecyclerView networkStatsRecycler;
    private SwipeRefreshLayout networkStatsRefreshLayout;
    private NetworkStatsAdapter adapter;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.network_stats_activity);
        networkStatsRecycler = findViewById(R.id.network_stats_recycler);
        networkStatsRefreshLayout = findViewById(R.id.network_stats_refresh_layout);
        networkStatsRefreshLayout.setOnRefreshListener(new SwipeRefreshLayout.OnRefreshListener() {

            @Override
            public void onRefresh() {
                presenter.onRefresh();
            }
        });
        networkStatsRefreshLayout.post(new Runnable() {
            @Override
            public void run() {
                networkStatsRefreshLayout.setRefreshing(true);
                presenter.onRefresh();
            }
        });
    }

    @Override
    protected NetworkStatsPresenter getPresenter() {
        return new NetworkStatsPresenterImpl(this);
    }

    @Override
    public void showNetworkStats(Collection<NetTrafficStats> netTrafficStats) {
        if (adapter == null) {
            adapter = new NetworkStatsAdapter(this, netTrafficStats, presenter);
            networkStatsRecycler.setLayoutManager(new LinearLayoutManager(this));
            networkStatsRecycler.setAdapter(adapter);
        } else {
            adapter.setDataList(netTrafficStats);
        }
    }

    @Override
    public void onRefreshCompleted() {
        networkStatsRefreshLayout.setRefreshing(false);
    }
}

File: app/src/main/java/com/zaze/demo/component/network/NetworkStatsActivity.kt
package com.zaze.demo.component.network

import android.app.AppOpsManager
import android.content.Context
import android.content.Intent
import android.os.Build
import android.os.Bundle
import android.provider.Settings
import androidx.databinding.DataBindingUtil
import androidx.lifecycle.Observer
import androidx.recyclerview.widget.LinearLayoutManager
import com.zaze.common.base.AbsActivity
import com.zaze.common.base.ext.obtainViewModel
import com.zaze.demo.R
import com.zaze.demo.databinding.NetworkStatsActBinding
import com.zaze.demo.debug.NetTrafficStats


/**
 * Description :
 *
 * @author : zaze
 * @version : 2017-12-22 04:36 1.0
 */
class NetworkStatsActivity : AbsActivity() {
    private var adapter: NetworkStatsAdapter? = null
    private lateinit var viewModel: NetworkStatsViewModel
    private lateinit var databinding: NetworkStatsActBinding


    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        databinding = DataBindingUtil.setContentView(this, R.layout.network_stats_act)
        databinding.lifecycleOwner = this
        databinding.viewModel = obtainViewModel(NetworkStatsViewModel::class.java).apply {
//            networkTraffic.observe(this@NetworkStatsActivity, Observer {
//                showNetworkStats(it)
//            })
            this@NetworkStatsActivity.viewModel = this
        }
        adapter = NetworkStatsAdapter(this, emptyList())
        databinding.networkStatsRecycler.layoutManager = LinearLayoutManager(this)
        databinding.networkStatsRecycler.adapter = adapter

        databinding.networkStatsBtn.setOnClickListener {
            hasPermissionToReadNetworkStats()
        }

        databinding.networkStatsRefreshLayout.setOnRefreshListener {
            viewModel.load()
        }
        if (hasPermissionToReadNetworkStats()) {
            databinding.networkStatsRefreshLayout.post {
                viewModel.load()
            }
        }
    }

    private fun hasPermissionToReadNetworkStats(): Boolean {
        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.M) {
            return true
        }
        val appOps = getSystemService(Context.APP_OPS_SERVICE) as AppOpsManager
        val mode = appOps.checkOpNoThrow(
            AppOpsManager.OPSTR_GET_USAGE_STATS,
            android.os.Process.myUid(), packageName
        )
        if (mode == AppOpsManager.MODE_ALLOWED) {
            return true
        }
        // 打开“有权查看使用情况的应用”页面
        val intent = Intent(Settings.ACTION_USAGE_ACCESS_SETTINGS)
        startActivity(intent)
        return false
    }

    private fun showNetworkStats(netTrafficStats: Collection<NetTrafficStats>?) {
        adapter?.setDataList(netTrafficStats) ?: let {
            adapter = NetworkStatsAdapter(this, netTrafficStats)
        }
    }

}